### PR TITLE
Add expiration date when set ticket in pending status

### PR DESCRIPTION
**Pull request by ddurieux**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #2246

This PR / feature add the possibility to set an expiration date when set a ticket in pending status + a field to add a reason (+ cron to pass status pending to assigned when expire + tests)

### CODE REVIEW

File: install/update_91_92.php [hunk 1]
```diff
@@ -798,6 +798,23 @@ function update91to92() {
       $DB->queryOrDie($query, "9.2 add saved searches alerts notification translation");
    }
 
+   // Manage pending status end date + comment
+   $migration->addField("glpi_tickets", "pendingenddate", "datetime");
```
Comment (trasher) [inline install/update_91_92.php]: It would be better to check for field existance, to handle future manual updates

File: tests/units/Entity.php [hunk 2]
```diff
@@ -78,4 +78,77 @@ public function testSonsAncestors() {
...
+   }
+
+   public function testPendingStatusOptionEnabled() {
+      $entity = new \Entity;
+      $e = getItemByTypeName('Entity', '_test_root_entity', true);
+      $input = [
+          'id' => $e,
+          'pendingenddate' => 0
+      ];
+      $entity->update($input);
```
Comment (trasher) [inline tests/units/Entity.php]: Test return ;)

File: tests/units/Entity.php [hunk 3]
```diff
@@ -78,4 +78,77 @@ public function testSonsAncestors() {
...
+   }
+
+   public function testPendingStatusOption2Days() {
+      $entity = new \Entity;
+      $e = getItemByTypeName('Entity', '_test_root_entity', true);
+      $input = [
+          'id' => $e,
+          'pendingenddate' => 2
+      ];
+      $entity->update($input);
```
Comment (trasher) [inline tests/units/Entity.php]: Test return

File: tests/units/Entity.php [hunk 4]
```diff
@@ -78,4 +78,77 @@ public function testSonsAncestors() {
...
+
+   public function testPendingStatusOptionParentValue() {
+      $entity = new \Entity;
+      // Disabled option
+      $e = getItemByTypeName('Entity', '_test_child_1', true);
+      $input = [
+          'id' => $e,
+          'pendingenddate' => \Entity::CONFIG_NEVER
+      ];
+      $entity->update($input);
```
Comment (trasher) [inline tests/units/Entity.php]: test return

File: tests/units/Entity.php [hunk 5]
```diff
@@ -78,4 +78,77 @@ public function testSonsAncestors() {
...
+
+      $val = $entity->getUsedConfig('pendingenddate', $e);
+      $this->variable($val)->isEqualTo(\Entity::CONFIG_NEVER);
+
+      // Enabled option
+      $input = [
+          'id' => $e,
+          'pendingenddate' => 0
+      ];
+      $entity->update($input);
```
Comment (trasher) [inline tests/units/Entity.php]: Test return

File: tests/units/Entity.php [hunk 6]
```diff
@@ -78,4 +78,77 @@ public function testSonsAncestors() {
...
+
+      $val = $entity->getUsedConfig('pendingenddate', $e);
+      $this->variable($val)->isEqualTo(0);
+
+      // 2 days option
+      $input = [
+          'id' => $e,
+          'pendingenddate' => 2
+      ];
+      $entity->update($input);
```
Comment (trasher) [inline tests/units/Entity.php]: Test return

File: tests/units/Entity.php [hunk 7]
```diff
@@ -78,4 +78,77 @@ public function testSonsAncestors() {
...
+
+      $val = $entity->getUsedConfig('pendingenddate', $e);
+      $this->variable($val)->isEqualTo(\Entity::CONFIG_NEVER);
+
+      // Enabled option
+      $input = [
+          'id' => $e,
+          'pendingenddate' => 0
```
Comment (trasher) [inline tests/units/Entity.php]: Same as for the harcoded '2'

File: tests/units/Entity.php [hunk 8]
```diff
@@ -78,4 +78,77 @@ public function testSonsAncestors() {
...
+      $val = $entity->getUsedConfig('pendingenddate', $e);
+      $this->variable($val)->isEqualTo(0);
+   }
+
+   public function testPendingStatusOption2Days() {
+      $entity = new \Entity;
+      $e = getItemByTypeName('Entity', '_test_root_entity', true);
+      $input = [
+          'id' => $e,
+          'pendingenddate' => 2
```
Comment (trasher) [inline tests/units/Entity.php]: Better to use a constant instead of harcoded '2'?

File: tests/units/Ticket.php [hunk 9]
```diff
@@ -1039,4 +1039,98 @@ public function testAssignAcl() {
...
+      $followup = new \TicketFollowup;
+
+      // First: Activate the option in the entity + activate followup
+      $e = getItemByTypeName('Entity', '_test_root_entity', true);
+      $input = [
+          'id' => $e,
+          'pendingenddate' => 0,
+          'pending_add_follow' => 1
+      ];
+      $entity->update($input);
```
Comment (trasher) [inline tests/units/Ticket.php]: test return

File: tests/units/Ticket.php [hunk 10]
```diff
@@ -1039,4 +1039,98 @@ public function testAssignAcl() {
...
+      // Create a ticket
+      $input = [
+          'name'           => 'test',
+          'content'        => 'test description',
+          'entities_id'    => $e,
+          'status'         => \Ticket::WAITING,
+          'pendingenddate' => date('Y-m-d H:i:s', (date('U') - 36000))
+      ];
+      $ticket_id = $ticket->add($input);
+      $this->integer($ticket_id);
```
Comment (trasher) [inline tests/units/Ticket.php]: Missing test; should be greater than 0

File: tests/units/Ticket.php [hunk 11]
```diff
@@ -1039,4 +1039,98 @@ public function testAssignAcl() {
...
+
+      $this->integer(count($followup->find("`tickets_id`='".$ticket_id."'")))->isEqualTo(1);
+
+      // go back ticket in assigned status manually
+      $input = [
+          'id'     => $ticket_id,
+          'status' => \Ticket::ASSIGNED
+      ];
+      $ticket->update($input);
```
Comment (trasher) [inline tests/units/Ticket.php]: test return

File: tests/units/Ticket.php [hunk 12]
```diff
@@ -1039,4 +1039,98 @@ public function testAssignAcl() {
...
+      $this->variable($ticket->fields['pendingenddate'])->isNull();
+      $this->variable($ticket->fields['pendingcomment'])->isNull();
+
+      // Return ticket status in pending state
+      $input = [
+          'id'     => $ticket_id,
+          'status' => \Ticket::WAITING,
+          'pendingenddate' => date('Y-m-d H:i:s', (date('U') - 36000))
+      ];
+      $ticket->update($input);
```
Comment (trasher) [inline tests/units/Ticket.php]: test return

File: tests/units/Ticket.php [hunk 13]
```diff
@@ -1039,4 +1039,98 @@ public function testAssignAcl() {
...
+      $cron   = new \CronTask;
+
+      // First: Activate the option in the entity
+      $e = getItemByTypeName('Entity', '_test_root_entity', true);
+      $input = [
+          'id' => $e,
+          'pendingenddate' => 0,
+          'pending_add_follow' => 0
+      ];
+      $entity->update($input);
```
Comment (trasher) [inline tests/units/Ticket.php]: test return

File: tests/units/Ticket.php [hunk 14]
```diff
@@ -1039,4 +1039,98 @@ public function testAssignAcl() {
...
+      // Create a ticket
+      $input = [
+          'name'           => 'test',
+          'content'        => 'test description',
+          'entities_id'    => $e,
+          'status'         => \Ticket::WAITING,
+          'pendingenddate' => date('Y-m-d H:i:s', (date('U') - 36000))
+      ];
+      $ticket_id = $ticket->add($input);
+      $this->integer($ticket_id);
```
Comment (trasher) [inline tests/units/Ticket.php]: greater than 0

File: tests/units/Ticket.php [hunk 15]
```diff
@@ -1039,4 +1039,98 @@ public function testAssignAcl() {
...
+
+      // Run the cron
+      // reset last run, in case we run test few minutes ago
+      $cron->getFromDBbyName('Ticket', 'expirePending');
+      $cron->resetDate();
```
Comment (trasher) [inline tests/units/Ticket.php]: test return?

File: install/update_91_92.php (hunk 1 above)
Comment (ddurieux) [inline install/update_91_92.php]: Ok, I will modify it ;)

File: tests/units/Entity.php (hunk 8 above)
Comment (ddurieux) [inline tests/units/Entity.php]: Not really possible because it's number of days, in the test I test with 2 days

Comment (trasher) [inline tests/units/Entity.php]: Oh, OK, no problem then :)

**Comment by wawax** [comment]
hi @ddurieux , should be close this PR or do you plan to update it ?

**Comment by trasher** [comment]
No changes nor feedbacks for months; I'm closing this one.

**Comment by cadamuroa** [comment]
Hi @ddurieux , any chance to apply in the future version? Thanl you!

**Comment by cconard96** [comment]
@cadamuroa There is a new pull request for pending status reasons, automatic followups, and expiration.
https://github.com/glpi-project/glpi/pull/8706

**Comment by cadamuroa** [comment]
Thank you very much for the information @cconard96 .

**Pull request closed without merging**